engine.lisp

```diff
@@ -148,9 +168,69 @@ Returns T if all constraints match, NIL otherwise."
   (loop for constraint in (cdr matcher)
         always (match-constraint constraint node)))
 
+(declaim (ftype (function (character list plump:node (function (plump:node) T))
+                          (values &optional null))
+                match-pair-depth))
+(defun match-pair-depth (combinator matcher parent extender)
+  "Match a combinator and matcher pair against a list of nodes.
+Returns a vector of matching nodes."
+  (declare (optimize speed))
+  (block match-pair-depth
+    (handler-case
+        (prog1 (values)
```

Comment:
The reason I did this was that SBCL was complaining about the mismatch in amount of return variables for the `case`. This could also be solved by using something like `nth-value`.